python/plugins/processing/algs/qgis/Aggregate.py

```diff
@@ -0,0 +1,273 @@
+                                      'TRUE',
+                                      field_def['delimiter']))
+            else:
+                expression = '{}({}, {})'.format(field_def['aggregate'],
+                                                 field_def['input'],
+                                                 group_by)
+            expr = self.createExpression(expression, da)
+            self.fields_expr.append(expr)
+        return True
+
+    def processAlgorithm(self, parameters, context, feedback):
+        expr_context = self.createExpressionContext(parameters, context)
+        self.group_by_expr.prepare(expr_context)
+        self.geometry_expr.prepare(expr_context)
+        for expr in self.fields_expr:
+            expr.prepare(expr_context)
+
+        source = self.source
+
+        # Group features in memory layers
+        feature = QgsFeature()
+        features = self.source.getFeatures()
+        count = self.source.featureCount()
+        if count:
+            progress_step = 50.0 / count
+        current = 0
+        groups = {}
+        keys = []  # We need deterministic order for the tests
+        for feature in features:
+            expr_context.setFeature(feature)
+            group_by_value = self.evaluateExpression(self.group_by_expr, expr_context)
+
+            # Get an hashable key for the dict
+            key = group_by_value
+            if isinstance(key, list):
+                key = tuple(key)
+
+            group = groups.get(key, None)
+            if group is None:
+                sink, id = QgsProcessingUtils.createFeatureSink(
+                    'memory:',
+                    context,
+                    source.fields(),
+                    source.wkbType(),
+                    source.sourceCrs())
+                layer = QgsProcessingUtils.mapLayerFromString(id, context)
+                group = {
+                    'sink': sink,
+                    'layer': layer,
+                    'feature': feature
+                }
+                groups[key] = group
+                keys.append(key)
+
+            group['sink'].addFeature(feature)
+
+            current += 1
+            feedback.setProgress(int(current * progress_step))
+            if feedback.isCanceled():
+                return
+
+        (sink, dest_id) = self.parameterAsSink(parameters,
+                                               self.OUTPUT,
+                                               context,
+                                               self.fields,
+                                               QgsWkbTypes.multiType(source.wkbType()),
+                                               source.sourceCrs())
+
+        # Calculate aggregates on memory layers
+        if len(keys):
+            progress_step = 50.0 / len(keys)
+        for current, key in enumerate(keys):
+            group = groups[key]
+            exp_context = group['layer'].createExpressionContext()
```

Comment:
Better to use self.createExpressionContext here, then add the layer scope afterward. Otherwise it'll miss the processing algorithm scope and correct project scope.